### ISSUE TITLE
How to propagate gradients after a assign op?

### ISSUE THREAD

**Issue by pnuzyf** [tensorflow/tensorflow#54568]
Hello,
I got the no gradient error if I changed the update of self.key_memory and self.long_memory to . assign. It seems that assign op would break gradient. If I used "=", then it will create a new tensor.

WARNING:tensorflow:Gradients do not exist for variables ['memory_layer_write_controller/hidden_layer_0/kernel:0', 'memory_layer_write_controller/hidden_layer_0/bias:0', 'memory_layer_write_controller/hidden_layer_1/kernel:0', 'memory_layer_write_controller/hidden_layer_1/bias:0', 'memory_layer_key_erase_layer/hidden_layer_0/kernel:0', 'memory_layer_key_erase_layer/hidden_layer_0/bias:0', 'memory_layer_key_add_layer/hidden_layer_0/kernel:0', 'memory_layer_key_add_layer/hidden_layer_0/bias:0', 'memory_layer_long_erase_layer/hidden_layer_0/kernel:0', 'memory_layer_long_erase_layer/hidden_layer_0/bias:0', 'memory_layer_long_add_layer/hidden_layer_0/kernel:0', 'memory_layer_long_add_layer/hidden_layer_0/bias:0'] when minimizing the loss.

class MemoryLayer(tf.keras.layers.Layer):
def init(self, controller_layers, controller_hidden_act='relu', controller_output_act=None, n_clusters=100, key_dims=8, long_dims=8, short_dims=8, temperature=0.1, alpha=0.1, is_short=False, name="memory_layer", **kwargs):
super(MemoryLayer, self).init(name=name, trainable=True, **kwargs)
self.controller_layers = controller_layers
self.controller_hidden_act = controller_hidden_act
self.controller_output_act = controller_output_act
self.n_clusters = n_clusters
self.key_dims = key_dims
self.long_dims = long_dims
self.short_dims = short_dims
self.temperature = temperature
self.alpha = alpha
self.is_short = is_short

    # self.key_memory = tf.compat.v1.get_variable(
    #     f"{self.name}_key_memory",
    #     shape=[self.n_clusters, self.key_dims],
    #     initializer=tf.compat.v1.glorot_normal_initializer())
    self.key_memory = self.add_weight(f"{self.name}_key_memory",
                          shape=(self.n_clusters, self.key_dims),
                          initializer=tf.keras.initializers.GlorotNormal(),
                          trainable=True)

    self.long_memory = self.add_weight(f"{self.name}_long_memory",
                          shape=(self.n_clusters, self.long_dims),
                          initializer=tf.keras.initializers.GlorotNormal(),
                          trainable=True)

    # self.long_memory = tf.compat.v1.get_variable(
    #     f"{self.name}_long_memory",
    #     shape=[self.n_clusters, self.long_dims],
    #     initializer=tf.compat.v1.glorot_normal_initializer())

    if is_short:
        self.short_memory = self.add_weight(f"{self.name}_short_memory",
                          shape=(self.n_clusters, self.short_dims),
                          initializer=tf.keras.initializers.GlorotNormal(),
                          trainable=False)

        self.short_erase_layer = DNNLayer([self.short_dims], 'sigmoid',
                                         name=f"{self.name}_short_erase_layer")  # short的删除层
        self.short_add_layer = DNNLayer([self.short_dims], 'tanh', name=f"{self.name}_short_add_layer")  #short的更新层

    self.write_controller = DNNLayer(controller_layers, controller_hidden_act, controller_output_act, name=f"{self.name}_write_controller") #write控制器

    self.read_controller = DNNLayer(controller_layers, controller_hidden_act, controller_output_act, name=f"{self.name}_read_controller") #read控制器

    self.key_erase_layer = DNNLayer([self.key_dims], 'sigmoid', name=f"{self.name}_key_erase_layer") #key的删除层
    self.key_add_layer = DNNLayer([self.key_dims], 'tanh', name=f"{self.name}_key_add_layer") #key的更新层

    self.long_erase_layer = DNNLayer([self.long_dims], 'sigmoid', name=f"{self.name}_long_erase_layer")  #long的删除层
    self.long_add_layer = DNNLayer([self.long_dims], 'tanh', name=f"{self.name}_long_add_layer")  # long的更新层

@tf.function
def call(self, inputs, training=None, **kwargs):
    attr_emb = inputs[0]  # [batch_size, attr_emb]
    long_emb = inputs[1]  # [batch, long_dims] 长期兴趣表征
    isActive = inputs[2]  # [batch_size, 1] 是否需要更新long、short memory

    active_status = tf.cast(tf.transpose(isActive, perm=[1, 0]), dtype='float32')

    denominator = tf.math.count_nonzero(isActive)  # 统计非0的个数

    # ＞0才更新
    if denominator > 0 and training:
        # write memory
        # [batch_size, key_dims]
        write_query = self.write_controller(tf.stop_gradient(attr_emb))  # 获取写控制器的表征

        # [batch, n_clusters]
        write_attention = compute_cosine_similarity(write_query, self.key_memory, temperature=self.temperature)
        #
        # #[1, n_clusters]
        write_memory_attention = tf.matmul(active_status, write_attention) / tf.cast(denominator, dtype=tf.float32)

        # [n_clusters, 1]
        write_memory_attention = tf.transpose(write_memory_attention, perm=[1, 0])

        # write key memory
        key_erase_vector = self.key_erase_layer(write_query)  # [batch, key_dims]
        key_erase_vector = tf.matmul(active_status, key_erase_vector) / tf.cast(denominator,
                                                                                dtype=tf.float32)  # [1,key_dims]
        #
        key_add_vector = self.key_add_layer(write_query)  # [batch, key_dims]
        key_add_vector = tf.matmul(active_status, key_add_vector) / tf.cast(denominator,
                                                                            dtype=tf.float32)  # [1,key_dims]

        key_val = tf.stop_gradient(self.key_memory) \
                          * (1. - self.alpha * tf.matmul(write_memory_attention, key_erase_vector)) \
                          + self.alpha * tf.matmul(write_memory_attention, key_add_vector)

        # self.key_memory.assign(key_val)
        tf.compat.v1.assign(self.key_memory, key_val)

        # write long_memory
        long_erase_vector = self.long_erase_layer(long_emb)  # [batch, long_dims]
        long_erase_vector = tf.matmul(active_status, long_erase_vector) / tf.cast(denominator,
                                                                                  dtype=tf.float32)  # [1,long_dims]
        #
        long_add_vector = self.long_add_layer(long_emb)  # [batch, long_dims]
        long_add_vector = tf.matmul(active_status, long_add_vector) / tf.cast(denominator,
                                                                              dtype=tf.float32)  # [1,long_dims]

        long_val = tf.stop_gradient(self.long_memory) \
                           * (1. - self.alpha * tf.matmul(write_memory_attention, long_erase_vector)) \
                           + self.alpha * tf.matmul(write_memory_attention, long_add_vector)

        # self.long_memory.assign(long_val)
        tf.compat.v1.assign(self.long_memory, long_val)

        # read memory
    # [batch, key_dims]
    read_query = self.read_controller(tf.stop_gradient(attr_emb))

    # [batch, n_clusters]
    read_attention = compute_cosine_similarity(read_query, self.key_memory, temperature=self.temperature)

    # read_attention = self.read_sim_layer([read_query, self.key_memory], temperature=self.temperature)

    long_cluster_emb = tf.matmul(read_attention, self.long_memory)  # [batch, long_dims] 长期兴趣表征

    return long_cluster_emb

**Comment by mohantym** [comment]
Hi @pnuzyf ! Could you please share same code as Colab gist as it will help us expedite the issue?
